src/platform_impl/linux/x11/event_processor.rs

```diff
@@ -1079,7 +1032,89 @@ impl<T: 'static> EventProcessor<T> {
+                                            text_with_all_modifiers,
+                                        },
+                                    },
+                                    is_synthetic: false,
+                                },
+                            });
+                        }
+                    }
+
                     ffi::XI_RawKeyPress | ffi::XI_RawKeyRelease => {
+                        // This is horrible, but I couldn't manage to respect keyboard layout changes
+                        // in any other way. In fact, getting this to work at all proved so frustrating
+                        // that I (@maroider) lost motivation to work on the keyboard event rework for
+                        // some months. Thankfully, @ArturKovacs offered to help debug the problem
+                        // over discord, and the following is the result of that debugging session.
+                        //
+                        // Without the XKB extension, the X.Org server sends us the `MappingNotify`
+                        // event when there's been a change in the keyboard layout. This stops
+                        // being the case when we select ourselves some XKB events with `XkbSelectEvents`
+                        // and the "core keyboard device (0x100)" (we haven't tried with any other
+                        // devices). We managed to reproduce this on both our machines.
+                        //
+                        // With the XKB extension active, it would seem like we're supposed to use the
+                        // `XkbStateNotify` event to detect keyboard layout changes, but the `group`
+                        // never changes value (it is always `0`). This worked for @ArturKovacs, but
+                        // not for me. We also tried to use the `group` given to us in keypress events,
+                        // but it remained constant there, too.
+                        //
+                        // We also tried to see if there was some other event that got fired when the
+                        // keyboard layout changed, and we found a mysterious event with the value
+                        // `85` (`0x55`). We couldn't find any reference to it in the X11 headers or
+                        // in the X.Org server source.
+                        //
+                        // `KeymapNotify` did briefly look interesting based purely on the name, but
+                        // it is only useful for checking what keys are pressed when we receive the
+                        // event.
+                        //
+                        // So instead of any vaguely reasonable approach, we get this: reloading the
+                        // keymap on *every* keypress. That's peak efficiency right there!
+                        //
+                        // FIXME: Someone please save our souls! Or at least our wasted CPU cycles.
```

Comment:
Have you measured the effects of this? Should this PR be blocked on https://github.com/rust-windowing/winit/issues/1634?

Reply:
I haven't measured the effect of this, no.

> Should this PR be blocked on #1634?

If it's enough of an issue to be a blocking one, then it should block the merger of the `new-keyboard` branch into `master`, rather than my `new-keboard-linux` branch's merging into `new-keyboard`. There's still some work to do after this PR gets merged, which would be more convenient to do on a branch with all the backend implementations rather than having to deal with a bunch of branches all at once.

Reply:
Note that even in latest master, keyboard events depend on raw events anyway. So, to resolve #1634 in X11 environment, we need to fix this logic not to depend on raw events regardless of this PR status.

Reply:
After reading #1634 a bit closer, the main issue raised seems to be that winit uses excessive amounts of CPU time when *in the background*. With that in mind, I don't really think #1634 is of much concern for this PR. Allowing keyboard events to be received while the application is in the background isn't something that's intentionally supported, IIRC. As such, untangling this PR's code from raw key events wouldn't be necessary (IMO) to resolve #1634.